### ISSUE TITLE
Cross provider defaults for KaaS

### ISSUE THREAD

**Issue by cornelius-keller** [giantswarm/roadmap#1513]

**Comment by nprokopic** [comment]
> and ideally in the future even have that managed by a single team

I would go even further and say that this is IMHO almost the only way this will work consistently and scale with number of providers and features.

**Comment by cornelius-keller** [comment]
I would suggest to agree on a provider interface which we can check in CI. This would work like a contract test. 
Maybe we can agree on a common json schema for all the providers and a github action that is checking it on every commit.
IMO we need contracts like this in order to scale accross team boundaries.

**Comment by alex-dabija** [comment]
> I would suggest to agree on a provider interface which we can check in CI. This would work like a contract test. Maybe we can agree on a common json schema for all the providers and a github action that is checking it on every commit. IMO we need contracts like this in order to scale accross team boundaries.

Yes, I agree. A schema to validate some of the required fields should not be a lot of effort, but how should we start enforcing it? I'm asking because I would prefer our builds to still be green and to plan for each setting.

Of course we could also go for the initial big bang. We enforce the current common settings and we have to work together to migrate the clusters (but this requires us to be aligned on our priorities).

Alternatively, we can have one team enforcing the settings and then have the clusters migrated for all providers. This has the advantage that we don't negotiate based on which team ends up with less work.

It's very important to me to keep builds green. I would like to see it as part of our way of working.

The technical problem is simple but the process one is a lot more complicated.

**Comment by puja108** [comment]
I had a related talk with @marians this week, where we touched upon Rainbow also needing to rely on a consistent interface, where if changes are planned/made these are communicated clearly as changes of contract and can be worked on between teams.

For fields in CRDs openapi schemas might work, but what Marian mentioned is that we sometimes also rely on labels and we need to cover that when we validate.

**Comment by alex-dabija** [comment]
Yes, we need to define our APIs: values.yaml, CRs, GS specific labels, GS specific annotations, etc. This is just the first step.

**Comment by cornelius-keller** [comment]
> It's very important to me to keep builds green. I would like to see it as part of our way of working.
:100: %, we can start with things that are common already like the oidc settings. then we can extend to for example network. Also we don't need to remove the current values immediately, we can mark them as deprecated and can introduce the new ones already. This might be a little more effort, but I think it's worth it.

**Comment by marians** [comment]
+1 to this effort! I suggest to adapt the issue title, as this is about more than defaulting.

I discussed this with Alex and Puja. Aligning the cluster app values schema across providers has potential to not only improve our user/developer experience but also simplify client implementation for cluster creation and modification a great deal. So the effort we invest I'm sure will be worth it.

**Comment by marians** [comment]
FYI I'm trying a few things with regard to visibility first. I'd like to be able to compare the different provider's value schemas, as a first step towards alignment.

**Comment by alex-dabija** [comment]
> Who is in? @cornelius-keller @alex-dabija @gawertm @Rotfuks

As mentioned today in the KaaS Sync, I'm up for it. So is @cornelius-keller .

**Comment by marians** [comment]
Update

- In the meantime I have had sessions with Alex and Nikola and then one with Cornelius.
- There are lots of discussion topics:
  - How to mix aligned and non-aligned (provider-specific) schema
  - What should be configurable and what not
  - How would we migrate to new schema
  - How exactly should our reference schema be structured?
- I learned that the process of alignment will be slow, iterative, and probably never-ending (sounds worse than it is, I guess)
- I have invited to a next call with Alex, Nikola, and Cornelius
- In the meantime I am thinking about how we can work with completely provider-specific schemas.

**Comment by marians** [comment]
FYI This RFC https://github.com/giantswarm/rfc/pull/54 by @pipo02mix is touching on the subject of configurability, in this case the kubernetes version provided with a cluster-app.

**Comment by bavarianbidi** [comment]
Also notable:
coming from [this PR thread](https://github.com/giantswarm/cluster-test-suites/pull/19#discussion_r1205079634) we've learned that https://github.com/giantswarm/capi-node-labeler-app is part of some CAPI (e.g. CAPA) clusters to label all the worker nodes. On other providers (CAPO, CAPV, CAPVCD, CAPZ) we don't make use of this app.

Once we know how a migration from Vintage to CAPI could look like, we could re-think if we need worker nodes labels as `worker` or if we could drop it.

cc @Rotfuks

**Comment by marians** [comment]
The capi-node-labeler-app also has no owner

**Comment by bavarianbidi** [comment]
> The capi-node-labeler-app also has no codeowner

we're already working on this [here](https://github.com/giantswarm/github/pull/920/files/bef4f03dbd1ecf245d029b1b2a993ed293ddec58..fbe38cb5568bb66385e4c13a009228ac673b6814?notification_referrer_id=NT_kwDOAUV0n7M2NjE5OTEyMDA0OjIxMzI5MDU1)

**Comment by Rotfuks** [comment]
Ticket currently dependent / waiting on https://github.com/giantswarm/giantswarm/issues/27020
